lib/bolt/config/transport/base.rb

```diff
@@ -0,0 +1,125 @@
+# frozen_string_literal: true
+
+require 'bolt/error'
+require 'bolt/util'
+
+module Bolt
+  class Config
+    module Transport
+      class Base
+        attr_reader :input
+
+        def initialize(data = {}, boltdir = nil)
+          assert_hash_or_config(data)
+          @input   = reference?(data) ? data : filter(data)
+          @config  = Bolt::Util.deep_merge(defaults, @input)
+          @boltdir = boltdir
+          validate
+        end
+
+        def [](key)
+          @config[key]
+        end
```

Comment:
Let's add `dig` as well, mostly for the sake of completeness. Should be simple:

```rb
def dig(*keys)
  @config.dig(*keys)
end
```

There are a few places in at least the tests where that would be helpful.